lib/galaxy/app.py

```diff
@@ -581,8 +586,30 @@ def __init__(self, configure_logging=True, use_converters=True, use_display_appl
 
         self._configure_tool_shed_registry()
         self._register_singleton(tool_shed_registry.Registry, self.tool_shed_registry)
-        # Tool Data Tables
+        #  Tool Data Tables
         self._configure_tool_data_tables(from_shed_config=False)
+        self._register_celery_galaxy_task_components()
+
+    def _register_celery_galaxy_task_components(self):
+        """
+        Register subtype class instance to support implementation of a user rate limit for execution of celery tasks.
+        The default supertype class does not enforce a user rate limit. This is the case if the celery_user_rate_limit
+        config param is the default value.
+        """
+        if self.config.celery_user_rate_limit:
+            task_before_start: GalaxyTaskBeforeStart
```

Comment:
I think we need to register `GalaxyTaskBeforeStart` regardless of whether `celery_user_rate_limit` is set or not: we refer to it in `before_start` in `GalaxyTask`, which is specified as the base class for tasks in the `galaxy_task` decorator. Unless I'm not reading this correctly?

Reply:
With lagom, the di container you are using, there is no need to register a type if there are no constructor parameters. It does this implicitly. So the code works correctly as is. I have integration tests for this. However, it is inefficient because it will create a new instance of GalaxyTaskBeforeStart everytime it is requested in the GalaxyTask.before_start method. So in order to make the code more efficient it is better to explicitly register GalaxyTaskBeforeStart as a singleton.